src/ray/object_manager/pull_manager.cc

```diff
@@ -122,57 +156,87 @@ void PullManager::UpdatePullsBasedOnAvailableMemory(size_t num_bytes_available)
+                     << " num bytes available: " << num_bytes_available_;
+      const auto last_request_it =
+          task_argument_bundles_.find(highest_task_req_id_being_pulled_);
+      RAY_CHECK(last_request_it != task_argument_bundles_.end());
+      DeactivatePullBundleRequest(task_argument_bundles_, last_request_it,
+                                  &highest_task_req_id_being_pulled_,
+                                  &object_ids_to_cancel);
     }
+  }
 
-    if (next_request_it == pull_request_bundles_.end()) {
-      // No requests in the queue.
+  // If we still have capacity after processing worker requests, try to
+  // activate task requests.
+  while (num_bytes_being_pulled_ < num_bytes_available_) {
+    if (!ActivateNextPullBundleRequest(task_argument_bundles_,
+                                       &highest_task_req_id_being_pulled_,
+                                       &objects_to_pull)) {
       break;
     }
+  }
 
-    if (next_request_it->second.num_object_sizes_missing > 0) {
-      // There is at least one object size missing. We should not activate the
-      // bundle, since it may put us over the available capacity.
+  // While we are over capacity, deactivate requests starting from the back of
+  // the task request queue.
+  while (num_bytes_being_pulled_ > num_bytes_available_) {
+    if (highest_task_req_id_being_pulled_ == 0) {
+      // No task requests are active.
       break;
     }
-
-    RAY_LOG(DEBUG) << "Activating request " << next_request_it->first
+    RAY_LOG(DEBUG) << "Deactivating task args request "
+                   << highest_task_req_id_being_pulled_
                    << " num bytes being pulled: " << num_bytes_being_pulled_
                    << " num bytes available: " << num_bytes_available_;
-    // There is another pull bundle request that we could try, and there is
-    // enough space. Activate the next pull bundle request in the queue.
-    if (!ActivateNextPullBundleRequest(next_request_it, &objects_to_pull)) {
-      // This pull bundle request could not be activated, due to lack of object
-      // size information. Wait until we have object size information before
-      // activating this pull bundle.
-      break;
-    }
+    const auto last_request_it =
+        task_argument_bundles_.find(highest_task_req_id_being_pulled_);
+    RAY_CHECK(last_request_it != task_argument_bundles_.end());
+    DeactivatePullBundleRequest(task_argument_bundles_, last_request_it,
+                                &highest_task_req_id_being_pulled_,
+                                &object_ids_to_cancel);
   }
 
-  std::unordered_set<ObjectID> object_ids_to_cancel;
-  // While the total bytes requested is over the available capacity, deactivate
-  // the last pull request, ordered by request ID.
+  // If we are still over capacity, deactivate requests starting from the back
+  // of the worker request queue.
   while (num_bytes_being_pulled_ > num_bytes_available_) {
```

Comment:
```suggestion
  while (num_bytes_being_pulled_ > num_bytes_available_ && highest_worker_req_id_being_pulled_ > 0) {
```

Consider this defensive change